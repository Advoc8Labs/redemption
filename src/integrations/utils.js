import { installPlaid } from './plaid';
import { emailLogin, installFirebase } from './firebase';
import { initializeTagManager } from './googleTagManager';

export function dynamicImport(dependency, async = true, index = 0) {
  if (async) {
    return new Promise((resolve) => {
      global.scriptjsLoadPromise.then(($script) => {
        $script.default(dependency, resolve);
      });
    });
  } else if (index < dependency.length) {
    return new Promise((resolve) => {
      global.scriptjsLoadPromise.then(($script) => {
        $script.default(dependency[index], resolve);
      });
    }).then(() => dynamicImport(dependency, false, index + 1));
  }
}

export async function initializeIntegrations(integrations, clientArg, init) {
  const gtm = integrations['google-tag-manager'];
  const fb = integrations.firebase;
  const plaid = integrations.plaid;

  const [, firebase] = await Promise.all([
    installPlaid(!!plaid),
    installFirebase(fb)
  ]);

  if (gtm) initializeTagManager(gtm);
  if (fb && !init) emailLogin(fb, firebase, clientArg);
}
