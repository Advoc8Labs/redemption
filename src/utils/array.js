/**
 * Inserts an element into a list without side effects.
 */
function justInsert(list, element, index, replace = true) {
    return [
        ...list.slice(0, index),
        element,
        ...list.slice(replace ? index + 1 : index)
    ];
}

/**
 * Removes an element from a list without side effects.
 */
function justRemove(list, index) {
    return [...list.slice(0, index), ...list.slice(index + 1)];
}

function toList(itemOrList) {
    return Array.isArray(itemOrList) ? itemOrList : [itemOrList];
}

export { justInsert, justRemove, toList };
