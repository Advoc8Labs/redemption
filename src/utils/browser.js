export function runningInClient() {
  return typeof window === 'object';
}
