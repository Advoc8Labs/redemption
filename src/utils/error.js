export class SdkKeyError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SdkKeyError';
    }
}

export class CompanyKeyError extends Error {
    constructor(message) {
        super(message);
        this.name = 'CompanyKeyError';
    }
}

export class UserKeyError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UserKeyError';
    }
}

export class FetchError extends Error {
    constructor(message) {
        super(message);
        this.name = 'FetchError';
    }
}
