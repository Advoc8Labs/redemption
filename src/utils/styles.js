import { adjustColor, alignmentMap } from './formHelperFunctions';
import Delta from 'quill-delta';

const bootstrapStyles = {
    padding: '0.375rem 0.75rem',
    boxSizing: 'border-box',
    transition: 'border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out',
    outline: 'none'
};

const mobileIndices = {
    gridColumnStart: 'mobile_column_index',
    gridColumnEnd: 'mobile_column_index_end',
    gridRowStart: 'mobile_row_index',
    gridRowEnd: 'mobile_row_index_end'
};

/**
 * Handles the translation of server-side properties into responsive CSS
 * attributes
 */
class ApplyStyles {
    constructor(element, targets) {
        this.element = element;
        this.styles = element.styles;
        this.mobileStyles = element.mobile_styles;
        this.targets = Object.fromEntries(targets.map((t) => [t, {}]));
        this.mobileTargets = Object.fromEntries(targets.map((t) => [t, {}]));
    }

    // Return CSS for the step dimensions / layout
    getLayout() {
        const layout = {
            gridColumnStart: this.element.column_index + 1,
            gridRowStart: this.element.row_index + 1,
            gridColumnEnd: this.element.column_index_end + 2,
            gridRowEnd: this.element.row_index_end + 2
        };

        if (
            Object.values(mobileIndices).some(
                (index) => !isNaN(parseInt(this.element[index]))
            )
        ) {
            const mobileLayout = {};
            Object.entries(mobileIndices).forEach(([key, i]) => {
                let mVal = parseInt(this.element[i]);
                if (isNaN(mVal)) mobileLayout[key] = layout[key];
                else {
                    if (['gridColumnEnd', 'gridRowEnd'].includes(key)) {
                        mVal += 2;
                    } else mVal++;
                    mobileLayout[key] = mVal;
                }
            });
            layout['@media (max-width: 478px)'] = mobileLayout;
        }

        return layout;
    }

    addTargets(...targets) {
        targets.forEach((target) => {
            this.targets[target] = {};
            this.mobileTargets[target] = {};
        });
    }

    // Return CSS for a particular target HTML element
    getTarget(target) {
        return {
            ...this.targets[target],
            '@media (max-width: 478px)': this.mobileTargets[target]
        };
    }

    setStyle(target, key, val) {
        this.targets[target][key] = val;
    }

    // Translate a set of server-side properties into CSS for a particular
    // target
    apply(target, properties, get) {
        // if not array, assume user passed in 1 element
        if (!Array.isArray(properties)) properties = [properties];

        const styles = properties.map((p) => this.styles[p]);
        this.targets[target] = { ...this.targets[target], ...get(...styles) };

        let mobileStyles = properties.map((p) => this.mobileStyles[p]);
        // If no mobile overrides, don't set breakpoint style
        if (mobileStyles.every((s) => s === undefined)) return;
        // Fall back to default style if a mobile style doesn't exist
        mobileStyles = properties.map((p) => {
            const ms = this.mobileStyles[p];
            return ms !== undefined ? ms : this.styles[p];
        });
        this.mobileTargets[target] = {
            ...this.mobileTargets[target],
            ...get(...mobileStyles)
        };
    }

    applyBorders(target, prefix = '', important = true) {
        // If color isn't defined on one of the sides, that means there's no border
        if (!this.styles[`${prefix}border_top_color`]) return;
        const i = prefix && important ? '!important' : '';
        this.apply(
            target,
            [
                `${prefix}border_top_color`,
                `${prefix}border_right_color`,
                `${prefix}border_bottom_color`,
                `${prefix}border_left_color`
            ],
            (a, b, c, d) => ({ borderColor: `#${a} #${b} #${c} #${d} ${i}` })
        );
        this.apply(
            target,
            [
                `${prefix}border_top_width`,
                `${prefix}border_right_width`,
                `${prefix}border_bottom_width`,
                `${prefix}border_left_width`
            ],
            (a, b, c, d) => ({
                borderWidth: `${a}px ${b}px ${c}px ${d}px ${i}`
            })
        );
        this.apply(
            target,
            [
                `${prefix}border_top_pattern`,
                `${prefix}border_right_pattern`,
                `${prefix}border_bottom_pattern`,
                `${prefix}border_left_pattern`
            ],
            (a, b, c, d) => ({ borderStyle: `${a} ${b} ${c} ${d} ${i}` })
        );
    }

    applySelectorStyles(target, prefix) {
        this.applyBorders(target, prefix);
        if (this.styles[`${prefix}background_color`]) {
            this.apply(target, `${prefix}background_color`, (a) => ({
                backgroundColor: `#${a} !important`
            }));
        }
        if (this.styles[`${prefix}font_color`]) {
            this.apply(target, `${prefix}font_color`, (a) => ({
                color: `#${a} !important`
            }));
        }
    }

    applyMargin(target) {
        this.apply(
            target,
            ['padding_top', 'padding_right', 'padding_bottom', 'padding_left'],
            (a, b, c, d) => ({
                margin: `${a}px ${b}px ${c}px ${d}px`
            })
        );
    }

    applyCorners(target) {
        this.apply(
            target,
            [
                'corner_top_left_radius',
                'corner_top_right_radius',
                'corner_bottom_right_radius',
                'corner_bottom_left_radius'
            ],
            (a, b, c, d) => ({
                borderRadius: `${a}px ${b}px ${c}px ${d}px`
            })
        );
    }

    applyBoxShadow(target) {
        this.apply(
            target,
            [
                'shadow_x_offset',
                'shadow_y_offset',
                'shadow_blur_radius',
                'shadow_color'
            ],
            (a, b, c, d) => ({
                boxShadow: `${a}px ${b}px ${c}px #${d}`
            })
        );
    }

    applyHeight(target) {
        this.apply(target, ['height', 'height_unit'], (a, b) => ({
            height: `${a}${b}`
        }));
    }

    applyWidth(target) {
        this.apply(target, ['width', 'width_unit'], (a, b) => ({
            width: `${a}${b}`
        }));
    }

    applyFontStyles(target, placeholder = false) {
        this.apply(target, 'font_weight', (a) => ({ fontWeight: a }));
        this.apply(target, 'font_family', (a) => ({ fontFamily: a }));
        this.apply(target, 'font_size', (a) => ({ fontSize: `${a}px` }));
        this.apply(
            target,
            placeholder ? 'placeholder_italic' : 'font_italic',
            (a) => ({ fontStyle: a ? 'italic' : 'normal' })
        );
        this.apply(
            target,
            placeholder ? 'placeholder_color' : 'font_color',
            (a) => ({ color: `#${a}` })
        );
    }

    applyRichFontStyles() {
        this.element.richFontStyles = {};
        const delta = new Delta(this.element.text_formatted);
        delta
            .filter((op) => op.insert)
            .forEach((op, i) => {
                const attrs = op.attributes;
                if (!attrs) return;

                const desktop = this._applyRichFontScreenStyles(attrs);
                if (attrs.start && attrs.end) desktop.cursor = 'pointer';
                const mobile = this._applyRichFontScreenStyles(
                    attrs,
                    'mobile_'
                );

                this.element.richFontStyles[i] = {
                    ...desktop,
                    '@media (max-width: 478px)': mobile
                };
            });
    }

    _applyRichFontScreenStyles(attrs, p = '') {
        const styles = {};

        if (attrs[`${p}size`]) styles.fontSize = `${attrs[`${p}size`]}px`;
        if (attrs[`${p}family`])
            styles.fontFamily = attrs[`${p}family`].replace(/"/g, "'");
        if (attrs[`${p}color`]) styles.color = `#${attrs[`${p}color`]}`;
        if (attrs[`${p}weight`]) styles.fontWeight = attrs[`${p}weight`];
        if (attrs[`${p}italic`]) styles.fontStyle = 'italic';

        const lines = [];
        if (attrs[`${p}strike`]) lines.push('line-through');
        if (attrs[`${p}underline`]) lines.push('underline');
        if (lines.length > 0) styles.textDecoration = lines.join(' ');

        return styles;
    }
}

function getImageStyles(element) {
    const as = new ApplyStyles(element, ['container', 'image']);
    as.apply('container', 'layout', (a) => ({
        justifyContent: a
    }));
    as.apply('container', 'vertical_layout', (a) => ({
        alignItems: a
    }));

    as.applyMargin('image');
    as.applyWidth('image');

    if (element.styles.line_height) {
        as.apply('text', 'line_height', (a) => ({
            lineHeight: `${a}px`
        }));
    }

    element.applyStyles = as;
}

function getProgressBarStyles(element) {
    const as = new ApplyStyles(element, ['container', 'barContainer', 'bar']);
    as.apply('container', 'vertical_layout', (a) => ({
        justifyContent: a
    }));
    as.apply('container', 'layout', (a) => ({
        alignItems: a
    }));
    as.applyFontStyles('container');

    as.apply('barContainer', 'vertical_layout', (a) => ({
        justifyContent: a
    }));
    as.apply('barContainer', 'layout', (a) => ({
        alignItems: a
    }));
    as.apply('barContainer', 'width', (a) => ({
        width: `${a}%`
    }));
    as.applyMargin('barContainer');

    as.apply('bar', 'bar_color', (a) => ({
        backgroundColor: `#${a}`
    }));

    element.applyStyles = as;
}

function getTextStyles(element) {
    const as = new ApplyStyles(element, ['container', 'text']);
    as.apply('container', 'layout', (a) => ({
        alignItems: alignmentMap[a],
        textAlign: a
    }));
    as.apply('container', 'vertical_layout', (a) => ({
        justifyContent: a
    }));
    if (element.styles.border_color) {
        as.apply('container', 'border_color', (a) => ({
            border: `1px solid #${a}`
        }));
    }
    as.applyMargin('text');
    if (element.styles.line_height) {
        as.apply('text', 'line_height', (a) => ({
            lineHeight: `${a}px`
        }));
    }
    as.applyRichFontStyles();

    element.applyStyles = as;
}

function getButtonStyles(element) {
    const as = new ApplyStyles(element, [
        'container',
        'button',
        'buttonActive',
        'buttonHover',
        'spinner'
    ]);

    as.apply('container', 'layout', (a) => ({
        alignItems: alignmentMap[a],
        textAlign: a
    }));
    as.apply('container', 'vertical_layout', (a) => ({
        justifyContent: a
    }));

    as.apply('button', 'background_color', (a) => ({
        backgroundColor: `#${a}`
    }));
    as.applyHeight('button');
    as.applyWidth('button');
    as.applyCorners('button');
    as.applyBorders('button');
    as.applyMargin('button');

    as.applyBorders('buttonHover', 'hover_');
    if (element.link !== 'none') {
        as.apply('buttonHover', 'background_color', (a) => {
            const color = `${adjustColor(a, -30)} !important`;
            return {
                backgroundColor: color,
                borderColor: color,
                transition: 'background 0.3s !important'
            };
        });
    }
    if (element.styles.hover_background_color) {
        as.apply('buttonHover', 'hover_background_color', (a) => ({
            backgroundColor: `#${a} !important`
        }));
    }

    as.applyBorders('buttonActive', 'selected_');
    if (element.styles.selected_background_color) {
        as.apply('buttonHover', 'selected_background_color', (a) => ({
            backgroundColor: `#${a} !important`
        }));
    }

    as.apply('spinner', 'show_spinner_on_submit', (a) => ({
        display: a ? 'default' : 'none'
    }));
    as.apply('spinner', ['height', 'height_unit'], (a, b) => {
        const thirdHeight = Math.round(a / 3);
        return {
            right: `-${a}${b}`,
            width: `${thirdHeight}${b}`,
            height: `${thirdHeight}${b}`
        };
    });
    as.applyRichFontStyles();

    element.applyStyles = as;
}

function getFieldStyles(field) {
    const targets = ['container', 'fc', 'field', 'error', 'active', 'hover'];
    const styles = new ApplyStyles(field, targets);

    styles.applyFontStyles('container');
    styles.apply('container', 'layout', (a) => ({
        alignItems: a
    }));
    styles.apply('container', 'vertical_layout', (a) => ({
        justifyContent: a
    }));
    styles.applyMargin('fc');
    styles.applySelectorStyles('active', 'selected_');
    styles.applySelectorStyles('hover', 'hover_');
    styles.apply('error', 'font_family', (a) => ({
        fontFamily: `#${a}`
    }));
    styles.apply('error', 'font_size', (a) => ({
        fontSize: `${a}px`
    }));

    const type = field.servar.type;
    switch (type) {
        case 'signature':
            styles.apply('fc', 'width', (a) => ({
                width: `${a}px`
            }));
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.applyCorners('field');
            styles.applyBorders('field');
            styles.applyBoxShadow('field');
            break;
        case 'rich_file_upload':
            styles.applyHeight('fc');
            styles.applyWidth('fc');
            styles.apply(
                'field',
                [
                    'cta_padding_top',
                    'cta_padding_right',
                    'cta_padding_bottom',
                    'cta_padding_left'
                ],
                (a, b, c, d) => ({
                    padding: `${a}px ${b}px ${c}px ${d}px`
                })
            );
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            break;
        case 'rich_multi_file_upload':
            styles.addTargets('ac', 'add');
            styles.applyHeight('field');
            styles.applyWidth('field');
            styles.applyHeight('ac');
            styles.applyWidth('ac');
            styles.apply(
                'ac',
                [
                    'uploader_padding_top',
                    'uploader_padding_right',
                    'uploader_padding_bottom',
                    'uploader_padding_left'
                ],
                (a, b, c, d) => ({
                    padding: `${a}px ${b}px ${c}px ${d}px`
                })
            );
            styles.apply(
                'add',
                [
                    'cta_padding_top',
                    'cta_padding_right',
                    'cta_padding_bottom',
                    'cta_padding_left'
                ],
                (a, b, c, d) => ({
                    padding: `${a}px ${b}px ${c}px ${d}px`
                })
            );
            styles.apply('add', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            break;
        case 'button_group':
            styles.apply('fc', 'layout', (a) => ({
                alignItems: `#${a}`
            }));
            styles.apply('fc', 'vertical_layout', (a) => ({
                justifyContent: a
            }));
            styles.applyHeight('field');
            styles.applyWidth('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.applyBoxShadow('field');
            styles.applyCorners('field');
            styles.applyBorders('field');
            styles.applyMargin('field');
            break;
        case 'checkbox':
            styles.applyMargin('fc');
            break;
        case 'dropdown':
        case 'gmap_state':
            styles.applyMargin('fc');
            styles.applyWidth('fc');
            styles.applyBorders('field');
            styles.applyCorners('field');
            styles.applyHeight('field');
            styles.applyBoxShadow('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            break;
        case 'pin_input':
            styles.applyMargin('fc');
            styles.applyWidth('field');
            styles.applyHeight('field');
            styles.applyBoxShadow('field');
            styles.applyCorners('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.apply('field', 'font_size', (a) => ({
                fontSize: `${a}px`
            }));
            styles.apply('field', 'font_color', (a) => ({
                color: `#${a}`
            }));
            break;
        case 'multiselect':
            styles.applyMargin('fc');
            styles.applyCorners('field');
            styles.applyBorders('field');
            styles.applyBoxShadow('field');
            styles.applyFontStyles('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.apply('field', 'font_size', (a) => ({
                height: `${parseInt(a) + 4}px`
            }));
            break;
        case 'select':
            styles.applyMargin('fc');
            styles.applyCorners('field');
            styles.applyBorders('field');
            styles.applyBoxShadow('field');
            styles.applyFontStyles('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.apply('field', 'font_size', (a) => ({
                height: `${parseInt(a) + 4}px`
            }));
            break;
        case 'hex_color':
            styles.applyMargin('fc');
            styles.applyCorners('field');
            styles.applyBorders('field');
            break;
        default:
            styles.addTargets(
                'sub-fc',
                'placeholder',
                'placeholderActive',
                'placeholderFocus',
                'tooltip'
            );
            styles.applyMargin('fc');
            styles.applyWidth('fc');
            styles.applyHeight('sub-fc');
            styles.applyBoxShadow('field');
            styles.applyCorners('field');
            styles.applyBorders('field');
            styles.applyFontStyles('field');
            styles.apply('field', 'background_color', (a) => ({
                backgroundColor: `#${a}`
            }));
            styles.applyFontStyles('placeholder', true);
            styles.apply('placeholder', 'font_size', (a) => ({
                lineHeight: `${a}px`
            }));
            if (type !== 'text_area') {
                styles.apply('placeholder', 'font_size', (a) => ({
                    marginTop: `-${a / 2}px`
                }));
            }
            if (field.styles.placeholder_transition === 'shrink_top') {
                styles.apply('placeholderFocus', 'font_size', (a) => {
                    const minFontSize = Math.min(a, 10);
                    return {
                        top: 0,
                        marginTop: `${minFontSize / 2}px`,
                        fontSize: `${minFontSize}px`
                    };
                });
                if (styles.selected_placeholder_color) {
                    styles.apply(
                        'placeholderActive',
                        'selected_placeholder_color',
                        (a) => ({
                            color: `#${a}`
                        })
                    );
                }
                styles.apply('field', ['height', 'height_unit'], (a, b) => ({
                    paddingTop: `${a / 3}${b}`
                }));
            } else {
                styles.setStyle('placeholderFocus', 'display', 'none');
            }

            styles.apply('tooltip', 'font_size', (a) => ({
                width: `${a}px`
            }));
            break;
    }

    field.applyStyles = styles;
}

function applyStepStyles(step) {
    step.images.forEach((e) => getImageStyles(e));
    step.texts.forEach((e) => getTextStyles(e));
    step.buttons.forEach((e) => getButtonStyles(e));
    step.servar_fields.forEach((e) => getFieldStyles(e));
    const pb = step.progress_bar;
    if (pb) getProgressBarStyles(pb);
    return step;
}

export {
    ApplyStyles,
    bootstrapStyles,
    applyStepStyles,
    getImageStyles,
    getTextStyles,
    getButtonStyles,
    getFieldStyles,
    getProgressBarStyles
};
