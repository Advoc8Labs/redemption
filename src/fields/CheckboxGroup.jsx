import React from 'react';
import ReactForm from 'react-bootstrap/Form';
import { getFieldValue } from '../utils/formHelperFunctions';
import { justInsert } from '../utils/array';

const handleCheckboxGroupChange = (
    e,
    servarKey,
    step,
    fieldValues,
    updateFieldValues
) => {
    const target = e.target;
    const opt = target.name;
    let newValues = null;
    step.servar_fields.forEach((field) => {
        const servar = field.servar;
        if (servar.key !== servarKey) return;

        const fieldValue = getFieldValue(field, fieldValues);
        const { value } = fieldValue;
        const newValue = target.checked
            ? [...value, opt]
            : value.filter((v) => v !== opt);
        if (fieldValue.repeated) {
            const { valueList, index } = fieldValue;
            newValues = updateFieldValues({
                [servar.key]: justInsert(valueList, newValue, index)
            });
        } else {
            newValues = updateFieldValues({ [servar.key]: newValue });
        }
    });
    return newValues;
};

function CheckboxGroup({
    field,
    fieldLabel,
    fieldVal,
    otherVal,
    step,
    fieldValues,
    updateFieldValues,
    fieldOnChange,
    handleOtherStateChange,
    onClick,
    selectCSS,
    hoverCSS
}) {
    const servar = field.servar;
    const otherChecked = fieldVal.includes(otherVal);
    return (
        <>
            {fieldLabel}
            {servar.metadata.options.map((opt) => {
                return (
                    <ReactForm.Check
                        type='checkbox'
                        id={`${servar.key}-${opt}`}
                        name={opt}
                        key={opt}
                        label={opt}
                        checked={fieldVal.includes(opt)}
                        onChange={(e) => {
                            const newValues = handleCheckboxGroupChange(
                                e,
                                servar.key,
                                step,
                                fieldValues,
                                updateFieldValues
                            );
                            fieldOnChange([servar.key], newValues);
                        }}
                        onClick={onClick}
                        style={{
                            display: 'flex',
                            alignItems: 'center',
                            marginBottom: '5px'
                        }}
                    />
                );
            })}
            {servar.metadata.other && (
                <div
                    style={{
                        display: 'flex',
                        alignItems: 'center'
                    }}
                >
                    <ReactForm.Check
                        type='checkbox'
                        id={`${servar.key}-`}
                        name={otherVal}
                        key={otherVal}
                        label='Other'
                        checked={otherChecked}
                        onChange={(e) => {
                            const newValues = handleCheckboxGroupChange(
                                e,
                                servar.key,
                                step,
                                fieldValues,
                                updateFieldValues
                            );
                            fieldOnChange([servar.key], newValues);
                        }}
                        onClick={onClick}
                        style={{
                            display: 'flex',
                            alignItems: 'center'
                        }}
                    />
                    <ReactForm.Control
                        type='text'
                        style={{
                            marginLeft: '5px',
                            height: `${parseInt(field.font_size) + 4}px`,
                            backgroundColor: `#${field.background_color}`,
                            border: `${field.border_width}px solid`,
                            borderColor: `#${field.border_top_color} #${field.border_right_color} #${field.border_bottom_color} #${field.border_left_color}`,
                            boxShadow: `${field.shadow_x_offset}px ${field.shadow_y_offset}px ${field.shadow_blur_radius}px #${field.shadow_color}`,
                            color: `#${field.font_color}`,
                            fontStyle: field.font_italic ? 'italic' : 'normal',
                            fontWeight: field.font_weight,
                            fontFamily: field.font_family,
                            fontSize: `${field.font_size}px`
                        }}
                        css={{
                            '&:focus': {
                                boxShadow: `${field.shadow_x_offset}px ${field.shadow_y_offset}px ${field.shadow_blur_radius}px #${field.shadow_color} !important`,
                                ...selectCSS
                            },
                            '&:hover': hoverCSS
                        }}
                        id={servar.key}
                        value={otherVal || ''}
                        onChange={(e) => {
                            const newValues = handleOtherStateChange(otherVal)(
                                e
                            );
                            fieldOnChange([servar.key], newValues);
                        }}
                        onClick={onClick}
                        maxLength={servar.max_length}
                        minLength={servar.min_length}
                        required={otherChecked}
                    />
                </div>
            )}
        </>
    );
}

export default CheckboxGroup;
