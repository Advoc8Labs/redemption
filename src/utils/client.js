import * as errors from './error';
import { initInfo } from './init';
import encodeGetParams from './string';

export default class Client {
    async begin(formKey) {
        const { userKey, apiKey } = initInfo();
        const params = encodeGetParams({
            flow_key: formKey,
            ...(userKey ? { fuser_key: userKey } : {})
        });
        const url = `https://api.feathery.tech/api/panel/step/?${params}`;
        const options = {
            cache: 'no-store',
            headers: { Authorization: 'Token ' + apiKey }
        };
        return fetch(url, options).then((response) => {
            const { status } = response;
            switch (status) {
                case 200:
                    return response.json();
                case 401:
                    return Promise.reject(
                        new errors.APIKeyError('Invalid API key')
                    );
                case 404:
                    return Promise.reject(
                        new errors.UserKeyError('Invalid User key')
                    );
                default:
                    return Promise.reject(
                        new errors.FetchError('Unknown error')
                    );
            }
        });
    }

    async submitStep(formKey, stepNum, servars, action) {
        // servars = [{key: <servarKey>, <type>: <value>}]
        const { userKey, apiKey } = initInfo();
        const url = `https://api.feathery.tech/api/panel/step/submit/`;
        const data = {
            ...(userKey ? { fuser_key: userKey } : {}),
            flow_key: formKey,
            step_number: stepNum,
            servars: action === 'next' ? servars : [],
            action
        };
        const options = {
            cache: 'no-store',
            headers: {
                Authorization: 'Token ' + apiKey,
                'Content-Type': 'application/json'
            },
            method: 'POST',
            body: JSON.stringify(data)
        };
        return fetch(url, options).then((response) => {
            const { status } = response;
            switch (status) {
                case 200:
                    return response.json();
                case 201:
                    return response.json();
                case 401:
                    return Promise.reject(
                        new errors.APIKeyError('Invalid API key')
                    );
                case 404:
                    return Promise.reject(
                        new errors.UserKeyError('Invalid User key')
                    );
                default:
                    return Promise.reject(
                        new errors.FetchError('Unknown error')
                    );
            }
        });
    }
}
