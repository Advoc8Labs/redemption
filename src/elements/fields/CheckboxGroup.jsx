import React from 'react';
import ReactForm from 'react-bootstrap/Form';
import { getFieldValue } from '../../utils/formHelperFunctions';
import { justInsert } from '../../utils/array';
import { bootstrapStyles } from '../../utils/styles';

const handleCheckboxGroupChange = (
    e,
    servarKey,
    step,
    fieldValues,
    updateFieldValues
) => {
    const target = e.target;
    const opt = target.name;
    let newValues = null;
    step.servar_fields.forEach((field) => {
        const servar = field.servar;
        if (servar.key !== servarKey) return;

        const fieldValue = getFieldValue(field, fieldValues);
        const { value } = fieldValue;
        const newValue = target.checked
            ? [...value, opt]
            : value.filter((v) => v !== opt);
        if (fieldValue.repeated) {
            const { valueList, index } = fieldValue;
            newValues = updateFieldValues({
                [servar.key]: justInsert(valueList, newValue, index)
            });
        } else {
            newValues = updateFieldValues({ [servar.key]: newValue });
        }
    });
    return newValues;
};

function CheckboxGroup({
    field,
    fieldLabel,
    fieldVal,
    otherVal,
    step,
    fieldValues,
    updateFieldValues,
    onChange,
    handleOtherStateChange,
    onClick
}) {
    const { servar, applyStyles } = field;
    const otherChecked = fieldVal.includes(otherVal);
    return (
        <div css={applyStyles.getTarget('fc')}>
            {fieldLabel}
            {servar.metadata.options.map((opt, i) => {
                return (
                    <ReactForm.Check
                        type='checkbox'
                        id={`${servar.key}-${i}`}
                        key={`${servar.key}-${i}`}
                        name={opt}
                        label={opt}
                        checked={fieldVal.includes(opt)}
                        onChange={(e) => {
                            const newValues = handleCheckboxGroupChange(
                                e,
                                servar.key,
                                step,
                                fieldValues,
                                updateFieldValues
                            );
                            onChange(newValues);
                        }}
                        onClick={onClick}
                        style={{
                            display: 'flex',
                            alignItems: 'center',
                            marginBottom: '5px'
                        }}
                        css={{
                            'input[type="checkbox"]': {
                                marginTop: 0,
                                marginBottom: 0
                            }
                        }}
                    />
                );
            })}
            {servar.metadata.other && (
                <div
                    style={{
                        display: 'flex',
                        alignItems: 'center'
                    }}
                >
                    <ReactForm.Check
                        type='checkbox'
                        id={`${servar.key}-`}
                        key={`${servar.key}-`}
                        name={otherVal}
                        label='Other'
                        checked={otherChecked}
                        onChange={(e) => {
                            const newValues = handleCheckboxGroupChange(
                                e,
                                servar.key,
                                step,
                                fieldValues,
                                updateFieldValues
                            );
                            onChange(newValues);
                        }}
                        onClick={onClick}
                        style={{
                            display: 'flex',
                            alignItems: 'center'
                        }}
                        css={{
                            'input[type="checkbox"]': {
                                marginTop: 0,
                                marginBottom: 0
                            }
                        }}
                    />
                    <ReactForm.Control
                        type='text'
                        css={{
                            marginLeft: '5px',
                            ...bootstrapStyles,
                            ...applyStyles.getTarget('field'),
                            '&:focus': applyStyles.getTarget('active'),
                            '&:hover': applyStyles.getTarget('hover')
                        }}
                        id={servar.key}
                        value={otherVal || ''}
                        onChange={(e) => {
                            const newValues = handleOtherStateChange(otherVal)(
                                e
                            );
                            onChange(newValues);
                        }}
                        onClick={onClick}
                        maxLength={servar.max_length}
                        minLength={servar.min_length}
                        required={otherChecked}
                    />
                </div>
            )}
        </div>
    );
}

export default CheckboxGroup;
