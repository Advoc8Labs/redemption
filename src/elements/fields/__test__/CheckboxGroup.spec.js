import React from 'react';
import { create, act } from 'react-test-renderer';
import CheckboxGroup from '../CheckboxGroup';
import { getFieldStyles } from '../../../utils/styles';

// import user from '@testing-library/user-event';
// import { render } from '@testing-library/react';

describe('CheckboxGroup', () => {
    it('renders a Radio Button group', () => {
        // Arrange
        const field = {
            id: 'fieldId',
            apply_styles: {},
            styles: {
                height: '50',
                height_unit: '%',
                width: '50',
                width_unit: 'px',
                background_color: '000000',
                uploader_padding_top: '0',
                uploader_padding_bottom: '0',
                uploader_padding_left: '0',
                uploader_padding_right: '0',
                cta_padding_top: '0',
                cta_padding_bottom: '0',
                cta_padding_left: '0',
                cta_padding_right: '0'
            },
            mobile_styles: {},
            servar: {
                key: 'key',
                name: 'Upload file...',
                metadata: {
                    file_types: 'image/*',
                    options: []
                },
                type: 'select'
            },
            icon_url: 'https://feathery.tech'
        };
        const props = {
            field: field,
            fieldLabel: 'label',
            fieldVal: 'val',
            otherVal: 'otherVal',
            step: { servar_fields: [field] },
            fieldValues: { key: 'val' },
            updateFieldValues: { newKey: 'newVal' },
            onChange: jest.fn(),
            handleOtherStateChange: jest.fn(),
            onClick: jest.fn()
        };
        getFieldStyles(props.field);

        // Act
        let component;
        act(() => {
            component = create(<CheckboxGroup {...props} />);
        });

        // Assert
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
