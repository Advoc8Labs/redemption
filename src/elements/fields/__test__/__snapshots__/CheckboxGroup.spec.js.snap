// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup renders a Radio Button group 1`] = `
<div
  className="css-jra06k-CheckboxGroup"
>
  label
</div>
`;
